projectq/backends/_awsbraket/_awsbraket.py

```diff
@@ -0,0 +1,481 @@
+            return False
+        if self.device == 'SV1':
+            if isinstance(g, (R)) and get_control_count(cmd) == 1:
+                return True
+            if g in (Z, Y, X, Swap) and get_control_count(cmd) == 1:
+                return True
+            if g == X and get_control_count(cmd) == 2:
+                return True
+            if get_control_count(cmd) == 0:
+                if isinstance(g, (R, Rx, Ry, Rz)):
+                    # TODO: add MatrixGate to cover the unitary operation
+                    return True
+                if g in (X, Y, Z, H, S, T, Sdag, Tdag, SqrtX, Swap):
+                    return True
+            if g in (Measure, Allocate, Deallocate, Barrier):
+                return True
+            return False
+        return False
+
+    def _reset(self):
+        """ Reset all temporary variables (after flush gate). """
+        self._clear = True
+        self._measured_ids = []
+
+    def _store(self, cmd):
+        """
+        Temporarily store the command cmd.
+
+        Translates the command and stores it in a local variable (self._cmds).
+
+        Args:
+            cmd: Command to store
+        """
+        if self._clear:
+            self._probabilities = dict()
+            self._clear = False
+            self._circuit = ""
+            self._allocated_qubits = set()
+
+        # Previous to store the gate, checks availability against the device
+        available = self.is_available(cmd)
+        if not available:
+            raise Exception('Invalid command: ' + str(cmd) +
+                            '. Please check the available commands'
+                            'for the device ' +
+                            self.device + '.')
+
+        gate = cmd.gate
+        if gate == Allocate:
+            self._allocated_qubits.add(cmd.qubits[0][0].id)
+            return
+        if gate == Deallocate:
+            return
+        if gate == Measure:
+            assert len(cmd.qubits) == 1 and len(cmd.qubits[0]) == 1
+            qb_id = cmd.qubits[0][0].id
+            logical_id = None
+            for tag in cmd.tags:
+                if isinstance(tag, LogicalQubitIDTag):
+                    logical_id = tag.logical_qubit_id
+                    break
+            # assert logical_id is not None
+            if logical_id is None:
+                logical_id = qb_id
+                self._mapper.append(qb_id)
+            self._measured_ids += [logical_id]
+            return
+        # This should work for all the devices
+        if isinstance(gate, (R)) and get_control_count(cmd) == 1:
+            ctrl_pos = cmd.control_qubits[0].id
+            qb_pos = cmd.qubits[0][0].id
+            angle = gate.angle
+            self._circuit += '{"control": ' + str(ctrl_pos) + \
+                             ', "target": ' + str(qb_pos) + \
+                             ', "angle": ' + str(angle) + \
+                             ', "type": "cphaseshift"}, '
```

Comment:
I'm sorry but I don't understand the proposal. This multiplication is inside a string so you are multiplying the string ', "type": "c"' num_control times. Other possibility is to separate the "c" but I think that this could make the string too complicated just for two cases

Reply:
My global idea would be to refactor the code in order to avoid having special cases for every situation and try as much as possible to have a general approach that we apply to a few specific cases rather than the other way around.
By applying my suggestion, it might become possible to avoid (except for a few exceptions like 'X' and 'CNOT') to have a code that does not need to have special cases for the case without, with 1 or more control qubits.

In principle, if the `is_available` method is coded properly, we know that at this point in the code, we should not receive any unsupported gate.

Making this code more general would make the code much easier to maintain in the future if AWS Braket supports new operations.

Reply:
After thinking about this, and because you have to take care not only of the num of controls, but of the gate angle if applicable, of several targets if applicable and if the gate is a "instanceof" or not, I think it is difficult to make a general approximation. It could be possible if a helper function is created and all the complexity is moved to this helper function. But IMHO that would be to make the "problem" from one place to another. 

If you prefer to go for a helper function, please let me know.

Reply:
I am not sure I understand your point in "moving the problem from one place to another"; at the moment, you have a lot of code duplication within the `_store` method. Also, I am not necessarily requiring you to introduce a helper method.

What about something like this:
```python
    gate = cmd.gate
    num_controls =  get_control_count(cmd)

    json_cmd = {
        'target': [qb.id for qureg in cmd.qubits for qb in qureg],
        'type': "c" * num_controls + _gationary[type(gate)]
    }

    # Special case for CNOT and Toffoli gates
    if isinstance(gate, XGate) and num_controls > 0:
        json_cmd['type'] = "c" * (num_controls-1) + 'cnot'

    if isinstance(gate, (R, Rx, Ry, Rz, Ph, Rxx, Ryy, Rzz)):
        json_cmd['angle'] = str(gate.angle)

    if num_controls:
        json_cmd['control'] = [qb.id for qb in cmd.control_qubits]

    self.circuit_ += json.dumps(json_cmd)
```

Reply:
Thanks a lot @Takishima 

I had to take into account that for 2 controls the name should be "controls" instead of "control" and for 2 targets it is "targets" instead of "target" and in addition the cases of Tdag and Sdag and some other detail... that made the thing just a little bit more elaborated

But your proposal was inspiring, thank you very much.

```
        json_cmd = { }

        if num_controls == 2:
            json_cmd['controls'] = [cmd.control_qubits[0].id, cmd.control_qubits[1].id]
        if num_controls == 1:
            json_cmd['control'] = cmd.control_qubits[0].id
        if isinstance(gate, (R, Rx, Ry, Rz)):
            json_cmd['angle'] = gate.angle

        if isinstance(gate, (SwapGate)):
            json_cmd['targets'] = [qb.id for qureg in cmd.qubits for qb in qureg]
        else:
            json_cmd['target'] = cmd.qubits[0][0].id

        if gate in (Sdag, Tdag):
            json_cmd['type'] = self._gationary[gate.__str__()]
        else:
            json_cmd['type'] = "c" * num_controls + self._gationary[gate_type]

        self._circuit += json.dumps(json_cmd) + ", "

```